docs/order-confirmation/index.md

```diff
@@ -0,0 +1,333 @@
+
+This is *variant type*[^1] named `t` with two *constructors*, `Sandwich` and
+`Burger`. In OCaml, it is customary for the primary type of a module to be
+called `t`.
+
+The `Item` module should contain modules that help us render an item, and to do
+that, we'll need functions that can return the price and the emoji[^2] for a
+given item. First, add the `toPrice` function:
+
+<<< Snippets.re#to-price
+
+If you decide to add, say, a hotdog, to the menu, you would need to:
+
+- Add a `Hotdog` constructor to `Order.t`
+- Add a `| Hotdog` branch to the switch expression of `Order.toPrice`
+
+Your OCaml code wouldn't compile if you just added a new constructor to
+`Item.t` without also updating `Item.toPrice`. Adding or removing constructors
+from a variant usually forces you to change relevant parts of your code.
+
+## `fun` sugar syntax
+
+Incidentally, there's a sugar syntax for functions whose entire body is a switch
+expression. It's called `fun`, and we can rewrite `Item.toPrice` to use it:
+
+<<< Snippets.re#to-price-fun
+
+We can also define `toEmoji` using the `fun` sugar syntax:
+
+<<< Snippets.re#to-emoji
+
+Using the `fun` syntax is completely equivalent to using a switch expression,
+so it's up to your personal taste whether you want to use this sugar syntax.
+
+## `Order.make`
+
+Now we're ready to define the `make` function which will render the `Item`
+component:
+
+<<< Snippets.re#make
+
+The `make` function has a single labeled argument, `~item`, of type `t`. This
+effectively means the `Item` component has a single prop named `item`.
+
+Note that this renders a single row of a table. We'll need another component
+to render a table containing all items in an order.
+
+## `Order` component
+
+Create a new file `src/order-confirmation/Order.re` and add the following code:
+
+<<< Snippets.re#order
+
+There's a lot going on here:
+
+- The primary type of `Order` is `array(Item.t)`, which is an array of
+  variants.
+- The `make` function has a single labeled argument, `~items`, of type `t`.
+  This means the `Order` component has a single prop named `items`.
+- We sum up the prices of all items using
+  [Js.Array.reduce](https://melange.re/v2.1.0/api/re/melange/Js/Array/index.html#val-reduce).
+  This is identical to JavaScript's [Array.reduce
+  method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce),
+  except that the Melange version requires the initial value to be passed in.
+- For each order, we render an `Item` component via
+  [Js.Array.map](https://melange.re/v2.1.0/api/re/melange/Js/Array/index.html#val-map),
+  which is equivalent to the [Array.map
+  method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map).
+- The `React.array` function has the type signature `array(React.element) =>
+  React.element`, meaning it converts an array of `React.element`s to
+  `React.element`. All expressions in JSX must have the type `React.element`.
```

Comment:
This is not accurate. JSX transforms to function calls, so the expressions can be really anything. Even ReasonReact is lenient in that sense, as uppercase components can take any type as children.

There are two reasons why `React.array` is needed here:
- We are rendering a `tbody` element, and all lowercase elements expect their children to be of type `React.element`
- We are "concatenating" the array of items with the "total" row

I am not sure how to explain this easily, maybe say something like:

> React allows to combine an array of children with a single element, like we are doing here with the array of items and the row that shows the total. But OCaml type system needs all elements in a collection to have the same type, so we convert the array of items to a `React.element` using `React.array`.

Reply:
I tried to tackle this by adding an entire section for this instead of relegating it to a single bullet point: https://github.com/melange-re/melange-for-react-devs/pull/9/commits/c85c07af3b7a875e9de643730cabce55df6dbf3e